tests/unit/test_charm.py

```diff
@@ -4,64 +4,309 @@
+        self.harness.charm._on_install("mock_event")
+
+        pgb_container = self.harness.model.unit.get_container(self._pgbouncer_container)
+
+        # Ensure pushing container config with no users copies users from userlist to pgbouncer.ini
+        self.harness.charm._push_userlist({"test1": "pw"})
+        initial_ini = pgb_container.pull(INI_PATH).read()
+        initial_userlist = pgb_container.pull(USERLIST_PATH).read()
+
+        self.harness.charm._push_container_config()
+        updated_userlist = pgb_container.pull(USERLIST_PATH).read()
+        self.assertEqual(initial_userlist, updated_userlist)
+
+        updated_ini = pgb_container.pull(INI_PATH).read()
+        self.assertNotEqual(initial_ini, updated_ini)
+        self.assertIn("test1", updated_ini)
+
+        # Pushing container config with user dict should update both config files.
+        self.harness.charm._push_container_config({"test2": "pw"})
+        updated_userlist2 = pgb_container.pull(USERLIST_PATH).read()
+        self.assertNotEqual(updated_userlist, updated_userlist2)
+        self.assertEqual(updated_userlist2, '"test2" "pw"')
+
+        updated_ini2 = pgb_container.pull(INI_PATH).read()
+        self.assertIn("test2", updated_ini2)
+
+        _reload_pgbouncer.assert_called()
+
+    @patch("charm.PgBouncerK8sCharm._reload_pgbouncer")
+    def test_push_pgbouncer_ini(self, _reload_pgbouncer):
+        self.harness.charm._push_userlist({"test-user": "pw"})
+        self.harness.charm._push_pgbouncer_ini(users=None, reload_pgbouncer=True)
+        _reload_pgbouncer.assert_called()
+
+        pgb_container = self.harness.model.unit.get_container(self._pgbouncer_container)
+        ini = pgb_container.pull(INI_PATH).read()
+        self.assertIn("test-user", ini)
+
+    @patch("charm.PgBouncerK8sCharm._reload_pgbouncer")
+    def test_push_userlist(self, _reload_pgbouncer):
+        self.harness.charm._push_userlist({"initial-user": "pw"})
+        pgb_container = self.harness.model.unit.get_container(self._pgbouncer_container)
+
+        self.harness.charm._push_userlist(users={"test-user": "pw"}, reload_pgbouncer=True)
+        _reload_pgbouncer.assert_called()
+
+        userlist = pgb_container.pull(USERLIST_PATH).read()
+        self.assertEqual('"test-user" "pw"', userlist)
+
+        self.harness.charm._push_userlist(users=None, reload_pgbouncer=True)
+
+    def test_get_userlist_from_container(self):
+        self.harness.update_config({"pgb_admin_users": "test-admin"})
+        self.harness.charm._push_container_config(users={"test-admin": "pass"})
+        users = self.harness.charm._get_userlist_from_container()
+        self.assertDictEqual(users, {"test-admin": "pass"})
+
+        self.harness.charm._push_container_config(users={})
+        empty_users = self.harness.charm._get_userlist_from_container()
+        self.assertDictEqual(empty_users, {})
+
+    def test_reload_pgbouncer(self):
+        pass
+
+    # UTILITIES
```

Comment:
Perhaps add `# ---------------` before and after

Reply:
Added, thanks for catching this.